cupy/fft/fft.py

```diff
@@ -172,6 +178,12 @@ def _fft(a, s, axes, norm, direction, value_type='C2C', overwrite_x=False,
         a = _exec_fft(a, direction, value_type, norm, axes[-1], overwrite_x,
                       out_size)
 
+    # TODO(leofang): ask for help
+    # For some reason, the assignment operator in _fft_c2c() (a = _exec_fft(
+    # ...)) silently switched the current device if multi-GPU cuFFT is used.
+    # This hotfix ensures we use whichever device we started with.
+    if config.use_multi_gpus:
+        cupy.cuda.runtime.setDevice(a.data.device_id)
```

Comment:
For reviewer(s): Please note this observation. I can't understand why would the assignment operation in the Python space touched the CUDA device setting...

Reply:
Does it cause also in `master` branch, or only in this branch?

Reply:
Only here as far as I can tell...

Reply:
Using multiple ways, I verified that this device switch did not happen in the Python space. Rather, it happened in the last stage of the multi-GPU cuFFT plan execution. The last device provided to `cufftXtSetGPUs()` was set enabled. So, this is a bug in cuFFT, not in this PR.

Reply:
Sorry for the noise. I removed an earlier comment made when the diagnosis wasn't fully completed, but the conclusion stands: ***This device switching bug is in cuFFT, not in this PR.*** 

With NVIDIA architect @maxpkatz's help, we confirm that when using multiple GPUs, after `cufftDestroy()` completes the current device is silently switched to the last one provided to `cufftXtSetGPUs()`. 

As a workaround, we can either use the hotfix here, or just cache the current device when entering `Plan1d.fft()`. I prefer the latter, what do you think @asi1024?

Reply:
The internal NVIDIA ticket number filed by @maxpkatz is 2761341.

Reply:
I took the liberty and moved the fix to `Plan1d`'s destructor in f3a876e.

Reply:
I feel that caching the current device in `Plan1d.__del__()` seems better too. Thanks for your investigation!

Reply:
These lines may be fixed after the ticket is resolved. Could you leave this comment as a TODO?
https://github.com/cupy/cupy/pull/2644/files#diff-0539f58816d3895bb08f76961618117eR391-R395

Reply:
We no longer need any hotfix in `cupy/fft/fft.py` after the commit f3a876e. Did you mean you want to label the linked comment as a TODO? I can certainly do this, but in the foreseeable future we'll need this hotfix, because the bug is present at least up to CUDA 10.1 Update 2 and very likely also in 10.2 (I don't have the environment to test). Even if NVIDIA fixes it in 10.3 or 11.0, we still need to support all earlier CUDAs until their EOF. So, instead of a TODO, I consider this as "won't fix"...

Reply:
> Even if NVIDIA fixes it in 10.3 or 11.0, we still need to support all earlier CUDAs until their EOF.
I see. I agree that we don't need to fix it. Thanks.